### ISSUE TITLE
Windows: Sign files for Smart App Control

### ISSUE THREAD

**Issue by bwoodsend** [pyinstaller/pyinstaller#6747]
### Discussed in https://github.com/pyinstaller/pyinstaller/discussions/6746

<div type='discussions-op-text'>

<sup>Originally posted by **Evernow** April 15, 2022</sup>
Windows is introducing a new features in Windows 11 called [Smart App Control](https://support.microsoft.com/en-gb/topic/what-is-smart-app-control-285ea03d-fa88-4d56-882e-6698afdb7003) later this year, it enforces code signing much more strictly than SmartScreen.

A major issue I encountered is that files Pyinstaller extracts for modules needed, such as pywintypes310.dll in my case may not be code signed, which would result in warnings such as these: 

![image](https://user-images.githubusercontent.com/44484725/163515872-2b978618-2340-4069-8c19-92e026dc698e.png)

Is there a way to potentially sign these files the same way we can sign already made pyinstaller executables? A way is to do `--onedir`, however it is not practical to send end users a folder zipped and have to instruct them on how to launch the application. 

Note: To currently test Smart App Control you need the latest W11 Insider ISO and then enable it in Windows Defender. You CANNOT simply update an older insider install or even less go from stable to insider, using Smart App Control requires a clean install of a build that comes with the feature. </div>

**Comment by bwoodsend** [comment]
Ughh, not Windows as well. I woke up about 10 minutes ago and my day is already ruined. We'll need to implement built in codesigning like we did for macOS (docs for which are [here](https://docs.microsoft.com/en-us/windows/win32/seccrypto/cryptography-tools)). According to [this page](https://docs.microsoft.com/en-us/windows/win32/seccrypto/cert2spc) and [that page](https://docs.microsoft.com/en-us/powershell/module/pki/new-selfsignedcertificate?view=windowsserver2022-ps), we should be able to create a to create dummy certificates for testing with - i.e. PyInstaller devs don't need to buy a certificate just to test this feature.

**Comment by rokm** [comment]
We should take this opportunity and rework the macOS signing as well, to perform signing on all collected files instead of just binaries. Often, shared libs end up collected as datas due to brute-force hooks, and those files will slip past the codesinging and end up breaking the application...

**Comment by Legorooj** [comment]
I'm happy to work on this at some point.

**Comment by RussNelson** [comment]
By the way, if you're running pyupdater, you don't get a chance to sign the executable, as it runs pyinstaller and the exe goes straight into the .zip file. Here's what I did in myfile.spec after the EXE finishes:

`import subprocess
subprocess.run("signtool -f cert.pfx -p PASSWORD -fd sha256 myfile.exe".split(" "))`

**Comment by cudmore** [comment]
Any progress on this? We are happy to pay for a Microsoft certificate and then use `signtool`.

1) Does anybody have a 3d party company to buy a certificate from? For example, we can buy a certificate from [digicert](https://www.digicert.com/order/order-1.php).

2) Does anybody have a detailed recipe for the process of signing the final exe with signtool?

We got our macOS app to be properly signed (as long as it is run from the Applications folder?). We really need to get our Windows exe signed. This is all for an NIH funded open-source software project [SanPy](https://cudmore.github.io/SanPy/).

It is rather depressing that all the devs of pyinstaller put some much work and expertise into making exe/app from Python to have them blocked by "virus protection". :(

**Comment by bwoodsend** [comment]
> Does anybody have a detailed recipe for the process of signing the final exe with signtool?

I can point you at [a github search](https://github.com/search?q=signtool+pyinstaller&type=code) with examples in it. Nothing more though. We've never done this. I did have a go at this when I opened this issue but after some terrible error messages and discovering just how badly Microsoft's documentation correlated with reality, I ran out of enthusiasm.

> It is rather depressing that all the devs of pyinstaller put some much work and expertise into making exe/app from Python to have them blocked by "virus protection". :(

We've never really confirmed that signing will do any good against dumb antiviruses – just vague heresy on the internet.

**Comment by Pjumpod** [comment]
my antivirus also report the exe file generate by pyinstaller is virus as well.
rather than disabled the antivirus, anyone have the solution?

**Comment by bwoodsend** [comment]
[Sigh] Report the false positive to your antivirus vendor.

And can we please not make this conversation about antiviruses...

**Comment by RussNelson** [comment]
I'm confused. Did y'all not see the solution I posted two years ago? You'll need a Microsoft code signing certificate.

`import subprocess
cs=["C:\\Program Files (x86)\\Windows Kits\\10\\App Certification Kit\\signtool.exe","sign", "-f","Code_Signing_Certificate.pfx","-p","ABCDEFGHIJKL","-fd","sha256", exe.name]
subprocess.run(cs, executable=cs[0])

coll = COLLECT(exe,
               a.binaries,
               a.zipfiles,
               a.datas, 
               strip=False,
               upx=True,
               upx_exclude=[],
               name='win')
`

**Comment by bwoodsend** [comment]
Do you not have to sign all the DLLs too?

**Comment by RussNelson** [comment]
Seemingly not. Windows 10 complains if the .EXE is unsigned, and is happy when it is signed.

**Comment by bwoodsend** [comment]
Hmm, in that case I could take a legitimate piece of Microsoft software, replace one of the DLLs with something evil and publish it and, as long as I keep the top level EXE intact, Smart Screen would tell the would-be victims that that the malicious application is certified trusted software published by Microsoft. I know that the word *Smart* in product names is only ever used ironically but that setup would be particularly not smart.

Adding that command to somewhere inside `EXE()` doesn't sound too bad but we'd still need to figure out how to generate a dummy certificate so that we can test the signing functionality. I don't want a feature this significant going in that we developers can't test or reproduce bugs for.

**Comment by rokm** [comment]
If we are going to integrate signing functionality on Windows, I will be pushing for signing all collected binaries, similarly to how we do it on macOS. 

Because if only exe needs to be signed, the user can sign it themselves post-hoc, regardless of onefile vs onedir mode (or as part of custom .spec file workflow, as in Russ' case). And they can do it with whatever custom parametrization of `signtool` they want. Whereas we will need to think hard about what to actually use (Reference the cert by filename? Assume it is in storage and reference it by name? Use timeserver or not? Etc.).

**Comment by RussNelson** [comment]
You signed for the certificate, you distributed the software, you think you can hide malware inside a DLL and not get caught?

The .spec file is a specification file. It specifies what kind of packaging should happen. I see nothing wrong with expecting people who want to sign their executables to have to modify the spec file by hand. The spec file is a program after all, and aren't we all programmers?

Just add my three-line nugget to the documentation, and let people sign however they want.

**Comment by rokm** [comment]
> You signed for the certificate, you distributed the software, you think you can hide malware inside a DLL and not get caught?

I think the scenario was: *you* sign for certificate and distribute your (legit) software, I take it and swap one of its DLLs with malicious one, and distribute the compromised bundle. If this is onedir build, the executable remains signed with your certificate, and the signature remains valid (unless there is also some sort of cabinet-style signing over contents going on). Whereas in onefile case, this would require modifying the embedded archive, and would invalidate signature on the executable.

**Comment by RussNelson** [comment]
You're probably correct, but that comes down to an inadequacy of an antivirus vendor, which @bwoodsend asked us not to make it about. We can jump through Microsoft's hoop, whether that hoop is sufficient or not isn't our concern.

**Comment by CyberFoxar** [comment]
As of now, it seems that Smart Screen does flag unsigned DLLs, at least from some empirical testing I'm conducting with my users.

Signing an exe made as `onefile` allows my user to run them without a hitch, but when signing _only_ the exe in `oneDir` mode (same authority, same certificate), my users get an annoying windows Smart Screen popup.

I'm testing various ways around this, as running the exe as `oneFile` lead to long startup times, whereas `oneDir` is pretty snappy. My current investigations are around signing a self-extracting zip, in the hopes that it'll cascade down, but I'm not sure it'll work.

Having a built-in way to code-sign all collected files in pyinstaller would remove a lot of headaches for sure.

**Comment by bwoodsend** [comment]
Are you able to tell if signing all the DLLs in onedir mode reduces startup time? Could you try with both the app in a random directory and with it moved into `C:\Program Files` (which supposedly gets special treatment)?

**Comment by CyberFoxar** [comment]
Signing DLLs does not affect startup time, that much is known. I was merely telling (as is known) that running in `oneDir` mode is faster than running in `oneFile` mode. 

My issue is that there's additional signing to be done to an app running in `oneDir` mode, whereas an app made as `oneFile` only needs the original exe signed.

As a sidenote, making a self-extracting archive with 7zip (sfx archive), signing the archive exe then extracting it seems to properly propagate signatures to all files. Or at least, smart screen stopped bothering my users after that's been done.

So one way to do it might be to have a "persistent extraction" mode, where you'd keep the files extracted from a `oneFile` run. This would enable people to sign their exe, whilst having the benefits coming from a `oneDir` run.

**Comment by rokm** [comment]
> As a sidenote, making a self-extracting archive with 7zip (sfx archive), signing the archive exe then extracting it seems to properly propagate signatures to all files. Or at least, smart screen stopped bothering my users after that's been done.

In that case, wouldn't a signed installer (i.e., install wizard) also do the trick?

**Comment by CyberFoxar** [comment]
> In that case, wouldn't a signed installer (i.e., install wizard) also do the trick?

Seems like it does ! However, pyinstaller is not currently capable of making an install wizard by itself, needing us to figure out other methods. Having a way to build a rudimentary install wizard via pyinstaller might be a valid workaround for code-signing apps that can't be bundled easily as one file.

**Comment by RussNelson** [comment]
I am using oneDir with a single executable. That executable is signed, then bundled into the .zip along with everything else. I have written a separate installer .exe (also signed) which is basically a fancy unzip which 1) downloads versions.gz, determines the current version, 2) downloads the appropriate .zip file named in the versions.gz file for the channel (selectable by a command line option -- there is no gui), 3) unzips it into the appropriate folder, and 4) creates a shortcut on the desktop.

With both of those files signed, Windows doesn't give me any trouble. The signing happens in the .spec file as I outlined earlier in this thread.

**Comment by CyberFoxar** [comment]
That's been exactly my experience too, except your way is more automated, but it basically boils down to this:
- build as `oneDir`
- sign the exe
- package in self-extracting zip, an "installer"
- sign the "installer" exe

It works, but a built-in solution, kinda like the mac signing protocol, would be ideal. One that you'd feed your `.pfx` or `.crt` + `.key` to, and it'll spit out a properly packaged solution.

Another issue I've seen with this approach, is that it's not "portable". Once you've extracted the files, you cannot move them to another machine without smart screen whining about it once again. You cannot, for example, have the extracted files sitting on a shared drive and have people copy them over to their machines. They _have_ to copy the "installer", extract it on their machine, and _then_ smart screen allows them to run.

One might argue that's what `oneFile` mode is for, but it's a limitation to be aware of.

I'd love to see a solution that would sign _all_ collected files in such a way that the directory could be copied and smart screen wouldn't complain. I, sadly, lack the time to test it manually at the moment, but this would be ideal. I think it might be possible, either through `signtool` or `osslsigncode` (linux software) to sign arbitrary binaries. That would involve signing _all_ the files inside the `_internal` folder, but it might be doable.

**Comment by Legorooj** [comment]
> However, pyinstaller is not currently capable of making an install wizard by itself

This is something I plan to work on in the next month or two, having a trivial installer (or even a non-trivial one) wouldn't actually be too complex.

**Comment by bwoodsend** [comment]
> Another issue I've seen with this approach, is that it's not "portable". Once you've extracted the files, you cannot move them to another machine without smart screen whining about it once again. You cannot, for example, have the extracted files sitting on a shared drive and have people copy them over to their machines. They have to copy the "installer", extract it on their machine, and then smart screen allows them to run.

To confirm, you're saying that signing the DLLs is required to keep Smart Screen happy when moving applications between machines?

**Comment by CyberFoxar** [comment]
> To confirm, you're saying that signing the DLLs is required to keep Smart Screen happy when moving applications between machines?

I _suspect_ this is the case. I've not yet had time to confirm it.
